GroClient add_data_series() improvement

nemozen opened the pull request:
- take time ranges into account when ranking available series
- refactor add_data_series to use a separate find_data_series function

nemozen committed:
refactor add_data_series to use a separate find_data_series function

no change in functionality, just separating the part that affects stored state

nemozen committed:
add time range before ranking in find_data_series

Default behavior remains the same (i.e. time range is the max
available). But if there's more than one series available that
satisfies the selection, then the time range can affect optimal
ranking so we should pass it in.

itsjohnward commented on api/client/gro_client.py:
```diff
@@ -126,9 +131,25 @@ def add_data_series(self, **kwargs):
                 len(data_series_list), entities))
+            # temporal coverage affects ranking so add time range if specified.
+            for data_series in data_series_list:
+                if kwargs.get('start_date'):
```
These dates are getting filtered out in `get_params_from_selection()`.

Also even if it wasn't, `dict((k + 's', v) for k, v in iter(list(get_params_from_selection(**series).items())))` would be converting the keys to `start_dates` and `end_dates` which wouldn't be valid inputs.

itsjohnward commented on api/client/gro_client.py:
```diff
@@ -126,9 +131,25 @@ def add_data_series(self, **kwargs):
+                    data_series['start_date'] = kwargs['start_date']
+                if kwargs.get('end_date'):
+                    data_series['end_date'] = kwargs['end_date']
             for data_series in self.rank_series_by_source(data_series_list):
```
Do note rank_series_by_source() still doesn't work as expected when passing in a full data series. https://grointelligence.atlassian.net/browse/CLEWS-17127

Only accepts metric/item/region/frequency. No source_id.

nemozen replied:
Right. Actuall it does accept (and correctly ignore) source_id but  the source_name remains in the output and mismatches. Fixed in lib.py

nemozen committed:
add support for start_date and end_date in get_params_from_selection and rank_series_by_source

nemozen committed:
filter out source_name in rank_series_by_source and correct docstring

Output source_id is different, so leaving input source_name is misleading
Docstring was incorrect on what the function yields

nemozen committed:
improve handling of http response code 204 in get_data

204 (No Content) is used when the request is valid but the answer is
empty (e.g. no data series or data points available for the given
selection). Unlike 3xx, we don't want to retry.

nemozen committed:
add growing_degree_days function to CropModel

nemozen committed:
in find_data_series(), yield all possible series  in order instead of just the highest ranked one

add_data_series() will still only use the first one, but this gives
more flexibility when using the lower level function find_data_series

nemozen committed:
Revert "in find_data_series(), yield all possible series  in order instead of just the highest ranked one"

This reverts commit 970aafbccd892845ad6847f8df879f7a7b6ed73d.

nemozen committed:
Revert "add growing_degree_days function to CropModel"

This reverts commit 08d17e8a7f0b64ec4cf9c1e9939d3addd8038d30.

tolothy reviewed (approved):
tested with 
`groClient.add_data_series(**{
                'metric': 'Area planted (area)',
                'item': 'Corn',
                'region': 'United States', 
                'start_date': '2019-01-01',
                'end_date': '2019-12-31',
})`

Prior to the change `{u'metric_id': 2580001, u'source_name': u'WASDE', u'region_id': 1215, u'end_date': u'2020-08-31T00:00:00.000Z', u'item_name': u'Corn', u'partner_region_name': u'World', u'frequency_id': 9, 'source_id': 106, u'region_name': u'United States', u'partner_region_id': 0, u'item_id': 274, u'start_date': u'2008-09-01T00:00:00.000Z', u'metric_name': u'Area Planted (area)'}` is added.

After the change `{'partner_region_id': 0, 'source_id': 25, 'item_id': 274, 'metric_name': 'Area Planted (area)', 'metric_id': 2580001, 'frequency_id': 9, 'partner_region_name': 'World', 'start_date': '2019-01-01', 'region_id': 1215, 'item_name': 'Corn', 'end_date': '2019-12-31', 'region_name': 'United States'}` is added, which means added start_date and end_date kwargs were taken into consideration when data series were ranked.
